pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go

```diff
@@ -25,3 +39,56 @@ func TestSuite(t *testing.T) {
 	s := &testSuite{}
 	suite.Run(t, s)
 }
+
+func createLockObject(holderIdentity string) (obj runtime.Object) {
+	objectMeta := metav1.ObjectMeta{
+		Namespace: leaderNamespace,
+		Name:      leaseName,
+		Annotations: map[string]string{
+			rl.LeaderElectionRecordAnnotationKey: fmt.Sprintf(`{"holderIdentity":"%s"}`, holderIdentity),
+		},
+	}
+	return &v1.ConfigMap{ObjectMeta: objectMeta}
+}
+
+func TestLeaderLeaseDurationExpiration(t *testing.T) {
+	c := &fakecorev1.FakeCoreV1{Fake: &core.Fake{}}
+
+	le := LeaderEngine{
+		LeaseName:       leaseName,
+		LeaderNamespace: leaderNamespace,
+		LeaseDuration:   1 * time.Second,
+		HolderIdentity:  "foo",
+		coreClient:      c,
+	}
+
+	c.AddReactor("get", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+		return true, createLockObject("foo"), nil
+	})
+
+	var err error
+	le.leaderElector, err = le.newElection(le.LeaseName, le.LeaderNamespace, le.LeaseDuration)
+	require.NoError(t, err)
+
+	le.EnsureLeaderElectionRuns()
+
+	assert.True(t, le.IsLeader())
+
+	// For whatever reason (like the leader losing connection to API server), simulate
+	// another agent becoming the leader
+	c.PrependReactor("get", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+		return true, createLockObject("bar"), nil
+	})
+
+	time.Sleep(2 * time.Second)
```

Comment:
We usually try not to put sleeps in the tests as it slows down the pipeline - Is this one really necessary ?

Reply:
Agreed, but sometimes that can be difficult if the interface doesn't allow for enough flexibility. The race detector doesn't like this test anyway so I'll need to try something else.